Examples/Detectors/Common/include/ActsExamples/DetectorCommons/DetectorBase.hpp

```diff
@@ -0,0 +1,111 @@
+// This file is part of the ACTS project.
+//
+// Copyright (C) 2016 CERN for the benefit of the ACTS project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+#include "Acts/Geometry/GeometryContext.hpp"
+
+#include <memory>
+#include <vector>
+
+namespace Acts {
+class GeometryContext;
+class TrackingGeometry;
+class DetectorElementBase;
+class IMaterialDecorator;
+class Logger;
+namespace Experimental {
+class Detector;
+}  // namespace Experimental
+}  // namespace Acts
+
+namespace ActsExamples {
+class IContextDecorator;
+struct Geant4ConstructionOptions;
+}  // namespace ActsExamples
+
+class G4VUserDetectorConstruction;
+
+namespace ActsExamples {
+
+/// Base class for detector instances
+class DetectorBase {
+ public:
+  DetectorBase();
+  DetectorBase(
+      std::vector<std::shared_ptr<const Acts::DetectorElementBase>>
+          detectorStore,
+      std::vector<std::shared_ptr<IContextDecorator>> contextDecorators);
+  virtual ~DetectorBase();
+
+  virtual const Acts::GeometryContext& geometryContext() const = 0;
+
+  virtual std::shared_ptr<const Acts::TrackingGeometry> trackingGeometry()
+      const = 0;
+  virtual std::shared_ptr<Acts::Experimental::Detector> gen2Geometry()
+      const = 0;
+  virtual std::vector<std::shared_ptr<IContextDecorator>> contextDecorators()
+      const;
+
+  /// Build the Geant4 detector construction
+  /// @note This throws an exception if Geant4 is not enabled
```

Comment:
I think you removed the exception now? Should we stick to an exception or return nullptr? After thinking about it I think the exception might the better idea.

Reply:
Good point - I thought I make it return `nullptr` similar to `gen2Geometry` but maybe I should just always throw and never return null